### ISSUE TITLE
Doesn't work with `MacroExpression` constants

### ISSUE THREAD

**Issue by Sija** [wyhaines/defined.cr#2]
Typical use-case is to populate `VERSION` constant from `shards version`, like below:

```cr
class Foo
  VERSION = {{ `shards version "#{__DIR__}"`.chomp.stringify }}
end
```

Doing so, creates a `MacroExpression` assigned to `VERSION`, which is not being handled by this shard.
Workaround btw, is to wrap the code in `{% begin %} ... {% end %}` block:

```cr
class Foo
  {% begin %}
    VERSION = {{ `shards version "#{__DIR__}"`.chomp.stringify }}
  {% end %}
end
```

Could this possibly be handled from this shard side, without a need to resort to such modifications in the codebase (which might be as well out of our control)?

**Comment by wyhaines** [comment]
That is an interesting limitation, and a relevant one, as this project is what powers a key part of the opentelemetry instrumentation work that I am doing.

I do not have an answer for you, yet, but I will see what I can figure out. Thank you for pointing it out.

**Comment by Sija** [comment]
@asterite Any ideas whether this could be solved at all? (:

**Comment by asterite** [comment]
I don't think the version should come from a command.

**Comment by asterite** [comment]
Another answer; I didn't look at how this shard works, so I don't know why this doesn't work, and I probably won't have time to look at it.

In any case, I would report a feature request or bug if there's a specific issue or request.

**Comment by wyhaines** [comment]
So....

It can be solved with a "workaround".

https://twitter.com/EmilyKager/status/1516946307650260992

The fundamental issue is that at the time that the macro is executed, the `VERSION` constant points at a MacroExpression. And unless I misunderstand the internals here (possible -- there is a lot to know about the internals, and most of it, I haven't learned yet), a MacroExpression is a node that encompasses a `{{ ... }}` or `{% ... %}` section that has not been executed/resolved yet. And there isn't any way in situ to force that.

That's where your observation that if it is wrapped in a `{% begin %}...{% end %}` block, it works, comes in. Because in that case, the enclosed MacroExpression has already been evaluated and resolved, so the constant is just pointing at the result of that MacroExpression by the time that it is checked.

Stuck.

I have a workaround. It's in a branch.

If there is anything better that could possibly work, it'd be lovely to know what it is, but what the workaround does is to reformulate the problem in a case where the constant points at a MacroExpression.

Basically, if that happens, then the macro writes a completely new version check. It locally sets a constant, using a generated name to avoid any chance of collisions, to the MacroExpression, wrapped in a `{% begin %}...{% end %}` block.

In the situation that you describe, @Sija, it works. If the MacroExpression were itself accessing `@type`, though, results would not be what is expected. It's definitely a workaround.

I am inclined to merge the branch, because given the intent of the shard, and of the version checks, and given that you ran into this problem yourself, I am sure that I or others will encounter the same issue, but this solution definitely falls into the category of a hack.

**Comment by Sija** [comment]
@wyhaines Your ~~hack~~, erm... workaround looks great! It solves 99% of the cases I've seen so far re: macro usage in `VERSION` constant. Might be worth mentioning this accessing `@type` limitation in the README though.

**Comment by wyhaines** [comment]
> I don't think the version should come from a command.

In general, I agree. I am sure that where something like this is used, though, it's an effort to reduce the number of places that the version is specified. Because YAML files can not import anything themselves, codebases often have the version repeated in at least two different places. One is the shard.yml file, and the other is in the VERSION constant that most Crystal projects define.

If VERSION itself gets defined from what is in the shard, then people only need to remember to maintain their version in a single place.

I have started to inject a standard spec into my projects which validates that the versions match between the constant and the shard, but it is no surprise that this approach of dynamically setting VERSION from the output of `shards version` is popping up.

**Comment by wyhaines** [comment]
> @wyhaines Your ~hack~, erm... workaround looks great! It solves 99% of the cases I've seen so far re: macro usage in `VERSION` constant. Might be worth mentioning this accessing `@type` limitation in the README though.

I am going to sit on this branch for a few hours, and then come back to it, but I will likely be merging it, updating the docs, and doing a new release a little later in my day.

**Comment by asterite** [comment]
Sorry, my relatively harsh response was because I saw this:

> Doing so, creates a MacroExpression assigned to VERSION, which is not being handled by this shard.

So it seemed like the problem was with this shard, so I didn't know why I was pinged to fix this.

But, apparently the problem is with `compare_versions`. The reduced version, something I recommend to report to the Crystal main repo, is this:

```crystal
VERSION = {{ echo "1.0.1" }}

{% if compare_versions(VERSION, "0.0.1") > 0 %}
  {% puts "YES" %}
{% else %}
  {% puts "NO" %}
{% end %}
```

It gives this error:

```
In foo.cr:1:11

 1 | VERSION = {{ echo "1.0.1" }}
               ^
Error: expected first argument to 'compare_versions' to be a StringLiteral, SymbolLiteral or MacroId, not MacroExpression
```

Maybe it should work, though it's not clear how. Probably the constant value must be expanded to be used by the version check. The question is: should the value be stored in the constant so that it's only executed once? Right now if you mention a constant in a macro, and it's only used in macros, it doesn't generate anything at runtime. But this would change that. But if we don't cache the value, the macro expression will be executed multiple times.

Maybe in this case it's not a big deal because the value of the constant is a string literal. That said, I'm not sure whether whenever you reference a constant, the constant's value must be cached.

**Comment by wyhaines** [comment]
Right. That is the reduced version of the issue, and I think @Sija pinged you because he was just wondering if there were something obvious that would make it work. I don't think that there is. :)

If compare_versions could somehow force the expansion of the MacroExpression, it would fix the issue, and the shard wouldn't have to resort to the kind of workaround that I found. As you say, though, making that work has a lot of considerations.

Thanks for hopping into the discussion, though. I certainly appreciate your time!

**Comment by wyhaines** [comment]
Version [0.3.3](https://github.com/wyhaines/defined.cr/releases/tag/v0.3.3) has been released.
